app/artifact-cas/internal/service/download.go

```diff
@@ -88,20 +91,44 @@ func (s *DownloadService) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	// Set headers
-	w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%s", info.FileName))
-	w.Header().Set("Content-Length", strconv.FormatInt(info.Size, 10))
+	s.log.Infow("msg", "download initialized", "digest", wantChecksum, "size", bytefmt.ByteSize(uint64(info.Size)))
 
-	s.log.Infow("msg", "download initialized", "digest", hash, "size", bytefmt.ByteSize(uint64(info.Size)))
+	gotChecksum := sha256.New()
+	// create temporary buffer to write to both the writer and the checksum
+	buf := bytes.NewBuffer(nil)
 
-	if err := b.Download(ctx, w, hash.Hex); err != nil {
+	// NOTE: we don't sent the file directly to the writer because we need to calculate the checksum
+	// and we want to send the file / even if partially only if the checksum matches
+	// this has a performance impact but it's the only way to ensure that the file is not corrupted
+	// and don't require client-side verification
+	mw := io.MultiWriter(buf, gotChecksum)
+	if err := b.Download(ctx, mw, wantChecksum.Hex); err != nil {
 		if errors.Is(err, context.Canceled) {
-			s.log.Infow("msg", "download canceled", "digest", hash)
+			s.log.Infow("msg", "download canceled", "digest", wantChecksum)
 			return
 		}
 
 		http.Error(w, sl.LogAndMaskErr(err, s.log).Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Verify the checksum
+	if got, want := fmt.Sprintf("%x", gotChecksum.Sum(nil)), wantChecksum.Hex; got != want {
+		msg := fmt.Sprintf("checksums mismatch: got: %s, want: %s", got, want)
```

Comment:
same here, we only push content to the browser if we know it matches the digest.